core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,176 @@ angular.module('oppia').factory('ImprovementModalService', [
+               * Iterates through the learner actions and expands the set of
+               * actions to be displayed. If all learner actions have been
+               * handled, no more expansion is possible.
+               */
+              $scope.expandActionsToRender = function() {
+                let i;
+                let previousStateName = null;
+                if (indexOfFirstDisplayedAction <
+                    playthrough.actions.length - 1) {
+                  let action = playthrough.actions[
+                    indexOfFirstDisplayedAction];
+                  previousStateName =
+                      action.actionCustomizationArgs.state_name.value;
+                }
+                for (i = indexOfFirstDisplayedAction - 1; i >= 0; i--) {
+                  let action = playthrough.actions[i];
+                  let currentStateName =
+                      action.actionCustomizationArgs.state_name.value;
+
+                  if (currentStateName !== previousStateName) {
+                    // When the latest state name and the state name of the
+                    // current learner action differ, they are part of
+                    // different contexts. So, we identify whether the size
+                    // of the block has crossed the threshold number of actions
+                    // per block. If it has, we don't add more learner actions
+                    // to the block.
+                    if (indexOfFirstDisplayedAction - i >= (
+                      MAX_UNRELATED_ACTIONS_PER_BLOCK)) {
+                      indexOfFirstDisplayedAction = i + 1;
+                      break;
+                    }
+                  }
+                  previousStateName = currentStateName;
+                }
+                // This is the case when all learner actions have been iterated
+                // over, and no more expansion is possible. This updation needs
+                // to happen because the indexOfFirstDisplayedAction would not
+                // be updated when the final display block has less than the
+                // threshold number of learner actions.
+                if (i === -1) {
+                  indexOfFirstDisplayedAction = 0;
+                }
+              };
+
+              /**
+               * Renders a table for MultipleIncorrectSubmissions issue for a
+               * more friendly UI.
+               */
+              $scope.renderIssueTable = function() {
+                var lars = LearnerActionRenderService;
+                var tableHtml =
+                  lars.renderFinalDisplayBlockForMISIssueHTML(
+                    playthrough.actions.slice(
+                      indexOfActionInFirstDisplayedBlock),
+                    indexOfActionInFirstDisplayedBlock + 1);
+                return tableHtml;
+              };
+
+              /**
+               * Returns the list of learner actions to currently be
+               * displayed in the playthrough modal.
```

Comment:
Could you also explain how the behaviour of this function differs for the MIS case? (Pretend the reader is not looking at the implementation.)

Reply:
Done.